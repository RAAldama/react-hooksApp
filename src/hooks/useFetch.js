import { useEffect, useRef, useState } from "react";

export const useFetch = (url) => {
    const [state, setState] = useState({data: null, loading: true, error: null});
    const isMounted = useRef(true);

    useEffect( () => {

        return () => {
            isMounted.current = false;
        }
    }, [])

    useEffect( () => {
        setState({data: null, loading: true, error: null})

        fetch(url).then(ans => 
            ans.json()
        ).then(data => {

            if(isMounted.current){
                setState({
                    data,
                    loading: false,
                    error: null,
                })
            }else{
                console.log('setSate no fue llamado');
            }

        }).catch( () => {
            setState({
                data: null,
                loading: false,
                error: 'No se pudo cargar la informaci√≥n'
            })
        })

    }, [url]);

    return state;
}
