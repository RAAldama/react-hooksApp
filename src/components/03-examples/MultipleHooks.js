import React from "react";
import { useCounter } from "../../hooks/useCounter";
import { useFetch } from "../../hooks/useFetch";
import "../02-useEffect/SimpleForm.css";

const MultipleHooks = () => {

  const {counter, increase} = useCounter(1);
  const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);
  
  //Se usa un ternario (CONDICION && SUCEDE ESTO)
  //La doble negaci√≥n (!!) convierte un null en falso
  const {author, quote} = !!data && data[0];

  return (
    <div>
      <h1>Custom Hooks!</h1>
      <h2>Breaking Bad quotes</h2>
      <hr />

      {
        loading ? (
            <div className="alert alert-info text-center">Loading...</div>
        ) : (
            <blockquote className="blockquote text-end">
                <p>{quote}</p>
                <footer className="blockquote-footer">{author}</footer>
            </blockquote>
        )
      }

      <button className="btn btn-primary" onClick={() => increase(1)}>
          Siguiente frase
      </button>

    </div>
  );
};

export default MultipleHooks;
