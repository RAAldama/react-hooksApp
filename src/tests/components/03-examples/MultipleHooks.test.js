import { shallow } from "enzyme"
import MultipleHooks from "../../../components/03-examples/MultipleHooks"
import { useCounter } from "../../../hooks/useCounter";
import { useFetch } from "../../../hooks/useFetch";

jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');

describe('Tests on <MultipleHooks />', () => {

    beforeEach( () =>  {
 
        useCounter.mockReturnValue({
            counter: 10,
            increase: () => {}
        });
 
    });

    test('should render correctly', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        });

        const wrapper = shallow(<MultipleHooks />);
        expect(wrapper).toMatchSnapshot();
    })

    test('should show the information correctly', () => {
        
        useFetch.mockReturnValue({
            data: [{
                author: 'Albus',
                quote: 'Hola Mundo'
            }],
            loading: false,
            error: null
        });

        const wrapper = shallow(<MultipleHooks />);
        expect( wrapper.find('.alert').exists() ).toBe(false);
        expect( wrapper.find('p').text().trim() ).toBe('Hola Mundo');
        expect( wrapper.find('footer').text().trim() ).toBe('Albus');

    })
    
})
