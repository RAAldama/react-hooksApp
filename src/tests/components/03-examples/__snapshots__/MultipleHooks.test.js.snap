// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <MultipleHooks /> should render correctly 1`] = `
<div>
  <h1>
    Custom Hooks!
  </h1>
  <h2>
    Breaking Bad quotes
  </h2>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</div>
`;
